python/tk_houdini/ui_generation.py

```diff
@@ -29,92 +29,336 @@ def __init__(self, engine, commands):
+        """This method returns a modified list of context commands.
+        
+        This is called directly as a part of the dynamic menu generation code
+        as houdini builds submenus when the user clicks on the top-level
+        Shotgun menu. This should execute quickly.
+        
+        """
+
+        # only need to do this once since the menu object doesn't persist
+        # across context switches. using the cached values is important since
+        # the dynamic menu is rebuilt on each click.
+        if not hasattr(self, '_context_commands'):
+
+            # get the registered commands, grouped in the usual way.
+            (context_cmds, cmds_by_app, favourite_cmds) = self._group_commands()
+
+            # ideally we'd mimic the static menu and have a context item 
+            # that contained a submenu with context-specific commands. this
+            # doesn't seem to be possible with the dynamic menus, so instead
+            # we create an extra command in the "Current Context" menu that 
+            # has the context name and, when clicked, opens the context in
+            # Shotgun in the browser. 
+            context_cmd = AppCommand(
+                name=self._get_context_name(),
+                command_dict={
+                 'properties': {
+                  'description': "Displays the current context, jumps to SG.",
+                  'type': "context_menu",
+                 },
+                 'callback': lambda: None,
+                },
+            )
+
+            # the context command won't be available in the callback map of
+            # the engine since it isn't registered. we don't want to register
+            # it because it is specific to the dynamic menu. 
+            self._context_menu_item_id = context_cmd.get_id()
+
+            # create the full list
+            cmds = [context_cmd]
+            cmds.extend(context_cmds)
+
+            self._engine.log_debug(
+                "Collected context commands for dynamic menu.")
+            self._context_commands = cmds
+
+        return self._context_commands
+
+    def _get_commands_by_app(self):
+        """This method returns a list of commands sorted by app name.
+        
+        This is called directly as a part of the dynamic menu generation code
+        as houdini builds submenus when the user clicks on the top-level
+        Shotgun menu. This should execute quickly.
+        
+        """
+
+        # only need to do this once since the menu object doesn't persist
+        # across context switches. using the cached values is important since
+        # the dynamic menu is rebuilt on each click.
+        if not hasattr(self, '_commands_by_app'):
+
+            (context_cmds, cmds_by_app, favourite_cmds) = self._group_commands()
+
+            cmds = favourite_cmds
+
+            # get the grouped commands, then flatten it out into a list
+            for app_name in sorted(cmds_by_app.keys()):
```

Comment:
Yeah, I tried my best to leave the menu ordering in place. Here's the [original commit](https://github.com/shotgunsoftware/tk-houdini/commit/1877268e2d867cc6d40b3ffe6a8f5cd36d5dd1be#diff-5493f20cabc79093f91df5321f1f2e10R97) with the sorting. I can take it out if that's preferable.